quinn-h3/src/body.rs

```diff
@@ -221,3 +225,86 @@ impl Stream for RecvBodyStream {
         }
     }
 }
+
+pub struct RecvBodyReader {
+    recv: FrameStream,
+    trailers: Option<HeadersFrame>,
+    conn: ConnectionRef,
+    stream_id: StreamId,
+    buf: Option<Bytes>,
+}
+
+impl RecvBodyReader {
+    pub(crate) fn new(recv: FrameStream, conn: ConnectionRef, stream_id: StreamId) -> Self {
+        Self {
+            recv,
+            conn,
+            stream_id,
+            trailers: None,
+            buf: None,
+        }
+    }
+
+    fn buf_pop(&mut self, buf: &mut [u8]) -> usize {
+        match self.buf {
```

Comment:
This looks like it could benefit from using `map_or()`, but maybe hold on that until I get my context straight and we figure out if `Option<Bytes>` is the best type for this data structure.

Reply:
Ah, missed that. 

`Option<Bytes>` is some only when the last polled payload hasn't been fully consumed (i.e the client buffer is smaller than the payload size). 

What I try to achieve here is to leverage the `Bytes::split_off()` behavior that only change a bunch of indexes from the original `Bytes` and refcounts the produced `Bytes` to keep data into a contiguous allocated buffer.

In the case where user's buffer is larger than the payload. `read()` will only pass this, and will not try to poll for another data frame. I thought it would maybe save some copies / allocations.

Reply:
It sounds to me like `buf_pop()` could actually just be called `read()`?

Reply:
Yeah, but as it would probably conflict with the `std::io::Read` impl, I'll rename it `buf_read()`. And maybe `buf_put()` would also be a better name for `buf_push()`.